### ISSUE TITLE
2 different faces but small distance?

### ISSUE THREAD

**Issue by huytq97** [luckycallor/InsightFace-tensorflow#15]
I test embeddings from get_embd.py with 2 different faces, but distance between embeddings is small?
link image:https://drive.google.com/drive/folders/1YGB7-gxwUUt-UB9pSfdvoNeOwE82zCF8?usp=sharing
tks

**Comment by Hamiltonsjtu** [comment]
Thanks for your contribution, great work!

Euclidian distance is so small, nealy $$10^{-7}$$. 

I do not know what is going wrong

**Comment by SharharZ** [comment]
I also have this problem, i try to compare 2 different faces in my data, the cosine similarity and distance were very similar. Whether the pre-trained model was error? but i evaluate the public dataset, the acc were high.

**Comment by luckycallor** [comment]
@huytq97 @Hamiltonsjtu @SharharZ 
Did you use the latest get_embd.py? There was a bug in get_embd.py, and I have fixed it two months ago. Please check with the latest code.

**Comment by huytq97** [comment]
@luckycallor 
I test embeddings from get_embd.py with 2 different faces and pretrained ms1m_100 step 1006k, , but distance between embeddings still is 1e-5, get_embd.py is latest
Can you check me out?
link 2 images test: https://drive.google.com/drive/folders/1Pcd7eTwmk2Uj3WmP0lhO_WgPiGlEfgl2?usp=sharing
Tks

**Comment by Hamiltonsjtu** [comment]
> 
> 
> @luckycallor
> I test embeddings from get_embd.py with 2 different faces and pretrained ms1m_100 step 1006k, , but distance between embeddings still is 1e-5, get_embd.py is latest
> Can you check me out?
> link 2 images test: https://drive.google.com/drive/folders/1Pcd7eTwmk2Uj3WmP0lhO_WgPiGlEfgl2?usp=sharing
> Tks

After checking the latest code with your uploaded faces, which come from different person. 
The similarity as following with the help of distance_metric = 1: 
 
face :0prefix_000013.jpg-->0prefix_000013.jpg similarity: 1.0 
face :8de0ed59-528d-11e9-8381-d8cb8a96a2010000040.png-->0prefix_000013.jpg similarity: 0.8464444279670715 
face :8de0ed59-528d-11e9-8381-d8cb8a96a2010000040.png-->8de0ed59-528d-11e9-8381-d8cb8a96a2010000040.png similarity: 1.0
![tick](https://user-images.githubusercontent.com/12877891/61933279-199be080-afb8-11e9-8b30-7f7c15f71896.png)

So, the latest embeds.py version tells the different faces are similar with amout 84.6%. 
I do know how to set the threshold to iditify a person. 
@luckycallor Could you give some commends for threshold?

**Comment by sealedtx** [comment]
@Hamiltonsjtu Hi, can you, please, share your code for generating embeddings and and calculating distance between two faces. Because I have faced same problem, my different face images have very small distance ~ 1.46304765e-05, I don't know where is mistake? Thank you!

**Comment by Hamiltonsjtu** [comment]
> 
> 
> @Hamiltonsjtu Hi, can you, please, share your code for generating embeddings and and calculating distance between two faces. Because I have faced same problem, my different face images have very small distance ~ 1.46304765e-05, I don't know where is mistake? Thank you!

Hi, sealedtx. 

I use the latest get_embd.py to generate embedding and take care the import functions.

**Comment by kscp123** [comment]
have you solved the problem? I also meet this problem. @sealedtx  @huytq97

**Comment by kscp123** [comment]
@Hamiltonsjtu Hi, can you provide more detials? I also use the last get_embd.py, but still face this problem. I just put two different faces into one folder and run the code to get embds_arr. Then calculate the cos similarity of embds_arr[0] and embds_arr[1] ,but the cos similarity is always 1. I try to solve it for a few days, thanks.

**Comment by Hamiltonsjtu** [comment]
> 
> 
> @Hamiltonsjtu Hi, can you provide more detials? I also use the last get_embd.py, but still face this problem. I just put two different faces into one folder and run the code to get embds_arr. Then calculate the cos similarity of embds_arr[0] and embds_arr[1] ,but the cos similarity is always 1. I try to solve it for a few days, thanks.

The script I used can be found https://github.com/Hamiltonsjtu/based_on_facenet/tree/master/InsightFace
as well as the images for test. And I do not change any script of the original project. Hope this script can help you find your bug. 

![distance](https://user-images.githubusercontent.com/12877891/64581580-4b78d300-d3bd-11e9-8fe5-16f3df600063.PNG)

**Comment by kscp123** [comment]
> > @Hamiltonsjtu Hi, can you provide more detials? I also use the last get_embd.py, but still face this problem. I just put two different faces into one folder and run the code to get embds_arr. Then calculate the cos similarity of embds_arr[0] and embds_arr[1] ,but the cos similarity is always 1. I try to solve it for a few days, thanks.
> 
> The script I used can be found https://github.com/Hamiltonsjtu/based_on_facenet/tree/master/InsightFace
> as well as the images for test. And I do not change any script of the original project. Hope this script can help you find your bug.
> 
> ![distance](https://user-images.githubusercontent.com/12877891/64581580-4b78d300-d3bd-11e9-8fe5-16f3df600063.PNG)

@Hamiltonsjtu Thanks for your help! I use your code and get better result(at least it's not always 1). It seems you change the way of reloading the pretrained model. So do you think the original loading way is wrong?

**Comment by Hamiltonsjtu** [comment]
> 
> 
> > > @Hamiltonsjtu Hi, can you provide more detials? I also use the last get_embd.py, but still face this problem. I just put two different faces into one folder and run the code to get embds_arr. Then calculate the cos similarity of embds_arr[0] and embds_arr[1] ,but the cos similarity is always 1. I try to solve it for a few days, thanks.
> > 
> > 
> > The script I used can be found https://github.com/Hamiltonsjtu/based_on_facenet/tree/master/InsightFace
> > as well as the images for test. And I do not change any script of the original project. Hope this script can help you find your bug.
> > ![distance](https://user-images.githubusercontent.com/12877891/64581580-4b78d300-d3bd-11e9-8fe5-16f3df600063.PNG)
> 
> @Hamiltonsjtu Thanks for your help! I use your code and get better result(at least it's not always 1). It seems you change the way of reloading the pretrained model. So do you think the original loading way is wrong?

I do not use yaml file frequently.  As you mentioned, I did a test by changing the Loader with the same images. 
Change 
`       config = yaml.load(open(args.config_path), Loader=yaml.FullLoader)`
into 
`        config = yaml.load(open(args.config_path))
`
At the end, I get the following results
![捕获](https://user-images.githubusercontent.com/12877891/64587701-c7314a80-d3d2-11e9-886e-fcd5308fb609.PNG)
which is different from the previous work. 
I do not know the reason, If you know, looking forward your reply. 
yaml documents can be found: [https://github.com/yaml/pyyaml/blob/master/lib3/yaml/loader.py](url)

**Comment by kscp123** [comment]
@Hamiltonsjtu 
Hi, I mean you use this way to restore pretrained model and it seems work.
saver = tf.train.import_meta_graph(args.model_path + '/best-m-200000.meta')
saver.restore(sess, args.model_path + '/best-m-200000')
But the author's original way gets wrong.
saver = tf.train.Saver(var_list=tf.trainable_variables())
saver.restore(sess, args.model_path)

For the use of yalm,  I have checked that after version 5.1, if Loader not given, it will be set FullLoader.

**Comment by sunruina2** [comment]
> > @Hamiltonsjtu Hi, can you provide more detials? I also use the last get_embd.py, but still face this problem. I just put two different faces into one folder and run the code to get embds_arr. Then calculate the cos similarity of embds_arr[0] and embds_arr[1] ,but the cos similarity is always 1. I try to solve it for a few days, thanks.
> 
> The script I used can be found https://github.com/Hamiltonsjtu/based_on_facenet/tree/master/InsightFace
> as well as the images for test. And I do not change any script of the original project. Hope this script can help you find your bug.
> 
> ![distance](https://user-images.githubusercontent.com/12877891/64581580-4b78d300-d3bd-11e9-8fe5-16f3df600063.PNG)

谢谢你分享的代码，在修改了模型reload方式之后，不同的图片会得到不同的embs，但却遇到了另外一个问题：同一张图片，运行get_embs两次的话，理论上两次embs的结果应该是一摸一样的，但是我却得到了不一样的结果，请问你有遇到类似的问题么？

**Comment by kscp123** [comment]
> > > @Hamiltonsjtu Hi, can you provide more detials? I also use the last get_embd.py, but still face this problem. I just put two different faces into one folder and run the code to get embds_arr. Then calculate the cos similarity of embds_arr[0] and embds_arr[1] ,but the cos similarity is always 1. I try to solve it for a few days, thanks.
> > 
> > 
> > The script I used can be found https://github.com/Hamiltonsjtu/based_on_facenet/tree/master/InsightFace
> > as well as the images for test. And I do not change any script of the original project. Hope this script can help you find your bug.
> > ![distance](https://user-images.githubusercontent.com/12877891/64581580-4b78d300-d3bd-11e9-8fe5-16f3df600063.PNG)
> 
> 谢谢你分享的代码，在修改了模型reload方式之后，不同的图片会得到不同的embs，但却遇到了另外一个问题：同一张图片，运行get_embs两次的话，理论上两次embs的结果应该是一摸一样的，但是我却得到了不一样的结果，请问你有遇到类似的问题么？

试一下这个reload方法
saver = tf.train.Saver()
saver.restore(sess, args.model_path)

**Comment by sunruina2** [comment]
> > > > @Hamiltonsjtu Hi, can you provide more detials? I also use the last get_embd.py, but still face this problem. I just put two different faces into one folder and run the code to get embds_arr. Then calculate the cos similarity of embds_arr[0] and embds_arr[1] ,but the cos similarity is always 1. I try to solve it for a few days, thanks.
> > > 
> > > 
> > > The script I used can be found https://github.com/Hamiltonsjtu/based_on_facenet/tree/master/InsightFace
> > > as well as the images for test. And I do not change any script of the original project. Hope this script can help you find your bug.
> > > ![distance](https://user-images.githubusercontent.com/12877891/64581580-4b78d300-d3bd-11e9-8fe5-16f3df600063.PNG)
> > 
> > 
> > 谢谢你分享的代码，在修改了模型reload方式之后，不同的图片会得到不同的embs，但却遇到了另外一个问题：同一张图片，运行get_embs两次的话，理论上两次embs的结果应该是一摸一样的，但是我却得到了不一样的结果，请问你有遇到类似的问题么？
> 
> 试一下这个reload方法
> saver = tf.train.Saver()
> saver.restore(sess, args.model_path)

@kscp123 可以了! 求教，这是什么原因呢？

**Comment by kscp123** [comment]
@sunruina2 应该是作者的方法只reload了可训练的参数，有一些模型里frozen的参数和BN的参数就被忽略了会导致错误。另外之前这位答主用的import_meta_graph方法我也不是很了解，了解的话请赐教

**Comment by sunruina2** [comment]
> @sunruina2 应该是作者的方法只reload了可训练的参数，有一些模型里frozen的参数和BN的参数就被忽略了会导致错误。另外之前这位答主用的import_meta_graph方法我也不是很了解，了解的话请赐教

@kscp123，做了4个对比测试供参考。主要原因是BN中的均值方差没有存在trainable_variables中，需要从global_variables取出来，再加入到trainable_variables里面，然后会得到正确结果了。和你执行的效果一样。(Four comparative tests have been done for reference. The main reason is that the mean and variance of BN do not exist in trainable_variables. It needs to be extracted from global_variables and added into trainable_variables, and then the correct results will be obtained. The effect is the same as that of your method.)

        # 1 @luckycallor version，Not available.
        # 不同人脸产出的emb向量相同
        # Different faces generate the same vectors. 
        saver = tf.train.Saver(var_list=tf.trainable_variables())
        saver.restore(self.sess, model_path)

        # 2 @Hamiltonsjtu version，Not available.
        # 不同人脸图片产出不同emb，但同一张图片跑两次产出不同emb，跑一次变一次。
        # Different faces generate the different vectors, but the same picture runs twice, produces two different embs, runs once and changes once. 
        saver = tf.train.import_meta_graph(model_path + '.meta')
        saver.restore(self.sess, model_path)


        # 3 @kscp123 version，available.
        # 不同人脸图片产出不同emb，同一张图片跑两次产出相同emb。可用。
        # Different faces generate the different vectors, and the same picture runs twice, produces same different embs. Available.
        saver = tf.train.Saver()
        saver.restore(self.sess, model_path)

        # 4 @sunruina2 version，available.
        # 运行结果同3
        # The operation effect is the same as 3
        var_list = tf.trainable_variables()
        g_list = tf.global_variables()
        bn_moving_vars = [g for g in g_list if 'moving_mean' in g.name]
        bn_moving_vars += [g for g in g_list if 'moving_variance' in g.name]
        # print(tf.trainable_variables(scope='embd_extractor/resnet_v2_50/block1/unit_2/block_v2/conv1/BatchNorm'))
        var_list += bn_moving_vars
        saver = tf.train.Saver(var_list=var_list)
        saver.restore(self.sess, model_path)
